### PR TITLE
INC launcher refinements

### PR DESCRIPTION
**Pull request by kaiyaointel**
# What does this PR do?

1. doc - add example doc for Neural Coder launcher for INC
2. doc - add launcher usage in homepage readme
3. default using Optimum API with Optimum launcher by setting ```use_inc=False```, this is echoing with this new [PR](https://github.com/intel/neural-compressor/pull/518) in INC repo
4. fix a bug (import path error): since we moved `launcher.py` file under `neural_compressor` per discussion from previous PR, the relative import path needs modified from `from .neural_compressor.neural_coder_adaptor import NeuralCoderAdaptor` to `from .neural_coder_adaptor import NeuralCoderAdaptor`

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by kaiyaointel** [comment]
@echarlaix Hi ella, please let us know if this PR makes sense to you :) Thanks!

File: examples/neural_compressor/launcher/PythonLauncher.md
Comment (echarlaix) [inline examples/neural_compressor/launcher/PythonLauncher.md:8]: I would be in favor of adding tests before giving visiblity to `NeuralCoderAdaptor`

File: README.md
Comment (echarlaix) [inline README.md:81]: We should remove the `eval_func` by default as it results in errors in for transformers examples

File: examples/neural_compressor/launcher/PythonLauncher.md
Comment (kaiyaointel) [reply examples/neural_compressor/launcher/PythonLauncher.md:8]: OK, let me add the tests as well in this PR later.

File: README.md
Comment (kaiyaointel) [reply README.md:81]: That's doable. Would you mind elaborating the "error" you mentioned? Is it a Optimum/INC algorithm-related issue, or does it related to this launcher?

**Comment by kaiyaointel** [comment]
Hi @echarlaix, I added the test for Neural Coder launcher, and the logic is that we will test if the optimized code of ```run_glue_source.py``` using this launcher is the same as the target file ```run_glue_target.py```, and thus can determine if the optimization with Neural Coder launcher is correct.

However, I saw that the github UT runs with official INC pypi version installation while this new UT here needs **the latest source code of INC** (https://github.com/intel/neural-compressor) because we enabled a new argument `use_inc` in Neural Coder's launcher in INC's source code in this [PR](https://github.com/intel/neural-compressor/pull/518) I mentioned (which, of course, is not yet in official INC pypi installation). This problem causes this UT to fail in this PR. How do we resolve this issue?

File: examples/neural_compressor/launcher/PythonLauncher.md
Comment (kaiyaointel) [reply examples/neural_compressor/launcher/PythonLauncher.md:8]: Test added, see comment below

File: README.md
Comment (echarlaix) [inline README.md:78]:
```suggestion
Dynamic quantization and ONNX export can be used through the Optimum Intel command-line:
```
(it would be great to have the ONNX export by default after `neural-compressor` next release)

**Review by echarlaix — COMMENTED** [review]
Thanks a lot for iterating on it @kaiyaointel !

File: README.md
Comment (kaiyaointel) [reply README.md:81]: Could you elaborate how do I do this? I'm not familiar with console_script entry. Thanks,

**Comment by kaiyaointel** [comment]
> Thanks a lot for iterating on it @kaiyaointel !

Thanks for the replies. See comments above. Regarding the error you mentioned (`IndentationError: unexpected indent`), would you try building INC from source code instead of installing the default pypi distribution of INC, because there're bug fixes in recent weeks in INC repo to fix this error.

File: README.md
Comment (echarlaix) [reply README.md:81]: We could add this directly in the setup.py file :
```python
setup(
    ...
    entry_points={"console_scripts": ["inc-quantization=optimum.intel.neural_compressor.launcher:main"]},
    ...
)
```
See te setuptools [documentation](https://setuptools.pypa.io/en/latest/userguide/entry_point.html). 
Also to be closer to the current optimum [CLI](https://huggingface.co/docs/optimum/main/en/onnxruntime/usage_guides/quantization#quantizing-a-model-to-be-used-with-optimums-cli), we can think of having something like:
```python
optimum-cli inc quantize --model distilbert-base-uncased-finetuned-sst-2-english quantized_distilbert/
```

File: tests/neural_compressor/run_glue_source.py
Comment (echarlaix) [inline tests/neural_compressor/run_glue_source.py:503]: Why not using the latest examples scripts of transformers by cloning the repo, and verifying that the model is correctly saved and can be loaded/used for inference, this would make the tests more robust in my opinion

Comment (kaiyaointel) [reply tests/neural_compressor/run_glue_source.py:503]: Sure, I can make this change. If we do this it means that we do not need `run_glue_source.py` and `run_glue_target.py` anymore right? Because we'll clone the latest example scripts on the fly every time the UT runs.

File: README.md
Comment (kaiyaointel) [reply README.md:81]: Thanks @echarlaix. Let me change the scripts according to your guidance.

Comment (kaiyaointel) [reply README.md]: I've modified the launcher to the CLI one `inc-quantization run_glue.py --model_name_or_path bert-base-cased --task_name mrpc --do_eval --output_dir result` pls refer to the latest change

File: tests/neural_compressor/run_glue_source.py
Comment (kaiyaointel) [reply tests/neural_compressor/run_glue_source.py]: @echarlaix Neural Coder default enables `save_onnx_model=False` in the line `quantizer.quantize(quantization_config=quantization_config, save_directory="quantized_model", save_onnx_model=False)`, because Neural Coder is designed for on-the-fly optimizing the code and then executes the optimized code directly (instead of saving the model) and thus it does not save the model by nature. Would it be okay if we simply tests whether the optimized code executes finely?

File: examples/neural_compressor/launcher/PythonLauncher.md
Comment (echarlaix) [reply examples/neural_compressor/launcher/PythonLauncher.md:8]: Thanks for adding it !

File: examples/neural_compressor/launcher/PythonLauncher.md
Comment (echarlaix) [inline examples/neural_compressor/launcher/PythonLauncher.md:14]: Already hard requirements of `optimum` so I would remove it 
```suggestion
```

File: README.md
Comment (echarlaix) [reply README.md]: I think having something like the following to match what we currently have in `optimum` could be great :
```python
optimum-intel-cli inc quantize --model distilbert-base-uncased-finetuned-sst-2-english quantized_distilbert/
```
What do you think ?

File: tests/neural_compressor/run_glue_source.py
Comment (echarlaix) [reply tests/neural_compressor/run_glue_source.py]: Thanks for iterating on it, it looks great ! The model will be saved after quantization in the `./quantized_model/` directory no ?

Comment (echarlaix) [reply tests/neural_compressor/run_glue_source.py]: Would love to have your opinion on adding a check to verify the model can be loaded (if that's feasible)

File: README.md
Comment (kaiyaointel) [reply README.md]: I think this makes sense. Let me modify to this design.

File: tests/neural_compressor/run_glue_source.py
Comment (kaiyaointel) [reply tests/neural_compressor/run_glue_source.py]: Sure, let me check

File: README.md
Comment (kaiyaointel) [reply README.md]: Have modified to this design. Please check latest changes. To use run_glue to quantize distilbert-sst2 using this CLI, the command is  ```optimum-intel-cli inc quantize run_glue.py --model_name_or_path distilbert-base-uncased-finetuned-sst-2-english --task_name sst2 --do_eval --output_dir result```

File: examples/neural_compressor/launcher/PythonLauncher.md
Comment (echarlaix) [inline examples/neural_compressor/launcher/PythonLauncher.md:4]:
```suggestion
Intel Neural Compressor quantization can be applied on your model through the Optimum Intel command-line:
```

File: optimum/intel/commands/inc/quantize.py
Comment (echarlaix) [inline optimum/intel/commands/inc/quantize.py:32]: Could this be moved to `optimum/intel/neural_compressor/.` (to have all the dependecy releated to INC there)

Comment (kaiyaointel) [reply optimum/intel/commands/inc/quantize.py:32]: I was designing completely based on your suggestion above - to mimic optimum's CLI, so that's why I created those files in those corresponding directory position. If you want our CLI to be `optimum-intel-cli inc quantize`, then those files have to be in those directories  (that's because `optimum-intel-cli` points to `optimum/intel/commands/`, so `optimum-intel-cli inc quantize` has to point to `optimum/intel/commands/inc/quantize')

File: tests/neural_compressor/run_glue_source.py
Comment (kaiyaointel) [reply tests/neural_compressor/run_glue_source.py]: test of loading saved model has been added to the UT

**Comment by kaiyaointel** [comment]
@echarlaix  Please take a final review on this PR.

- I have added the part of testing loading saved model `torch.load("./quantized_model/pytorch_model.bin")` to the UT file. Although, for automatic CI on UT, it might fail because it does not use latest INC repo build from source (it uses official INC pypi distribution).
- CLI `optimum-intel-cli inc quantize` coding is done, and I replied to your comments above regarding file location. It's completely same to optimum-cli design, so if we want to keep `optimum-intel-cli inc quantize` CLI design as you suggested, then the file location cannot be changed.

Let me know if any other comment.

File: optimum/intel/commands/inc/quantize.py
Comment (echarlaix) [reply optimum/intel/commands/inc/quantize.py:32]: Added a suggestion in https://github.com/kaiyaointel/optimum-intel/pull/3

**Comment by echarlaix** [comment]
> Although, for the current automatic CI on UT, it would fail because it does not use latest INC repo build from source (it uses official INC pypi distribution).

The tests fail because `neural-compressor` v2.1 is needed but also because the accuracy tolerance are not met, I would be in favor of increasing it by default so that it doesn't fail for users, what do you think ?


> Let me know if any other comment.

Just opened https://github.com/kaiyaointel/optimum-intel/pull/3 to include them.


Happy to merge https://github.com/huggingface/optimum-intel/pull/183 once the next release of `neural-compressor` is out.

**Comment by kaiyaointel** [comment]
> > Although, for the current automatic CI on UT, it would fail because it does not use latest INC repo build from source (it uses official INC pypi distribution).
> 
> The tests fail because `neural-compressor` v2.1 is needed but also because the accuracy tolerance are not met, I would be in favor of increasing it by default so that it doesn't fail for users, what do you think ?
> 
> > Let me know if any other comment.
> 
> Just opened [kaiyaointel#3](https://github.com/kaiyaointel/optimum-intel/pull/3) to include them.
> 
> Happy to merge #183 once the next release of `neural-compressor` is out.

Thanks much Ella for your revision. I have adopted the commit in this PR.

Let's merge this PR once the next release of `neural-compressor` is out.

**Review by echarlaix — APPROVED** [review]
Thanks a lot for integrating the suggestions @kaiyaointel !

**Comment by echarlaix** [comment]
Waiting for the tests to pass before merging, could you rebase @kaiyaointel ?

**Comment by kaiyaointel** [comment]
> Waiting for the tests to pass before merging, could you rebase @kaiyaointel ?

Sure, would you mind instructing on how to rebase? Also, I noticed that in the test, the failure seems to be pointing at `test_optimization.py` instead of the new UT added in this PR (`test_neural_coder_adaptor.py`), I wonder if it looked like the issue with INC new release.

**Comment by kaiyaointel** [comment]
> > Waiting for the tests to pass before merging, could you rebase @kaiyaointel ?
> 
> Sure. I have rebased.

Hi @echarlaix after rebasing I saw that all tests are passed. We can finally merge this PR :)

**Pull request merged**
